src/kinetics/InterfaceKinetics.cpp

```diff
@@ -610,4 +614,108 @@ double InterfaceKinetics::interfaceCurrent(const size_t iphase)
+    // reverse reaction rate coefficients
+    vector_fp& rop_rates = m_rbuf0;
+    processFwdRateCoefficients(rop_rates.data());
+    processEquilibriumConstants(rop_rates.data());
+    return calculateCompositionDerivatives(m_revProductStoich, rop_rates);
+}
+
+Eigen::SparseMatrix<double> InterfaceKinetics::netRatesOfProgress_ddCi()
+{
+    // check derivatives are valid
+    assertDerivativesValid("InterfaceKinetics::netRatesOfProgress_ddCi");
+    // forward reaction rate coefficients
+    vector_fp& rop_rates = m_rbuf0;
+    processFwdRateCoefficients(rop_rates.data());
+    Eigen::SparseMatrix<double> jac = calculateCompositionDerivatives(m_reactantStoich,
+        rop_rates);
+
+    // reverse reaction rate coefficients
+    processEquilibriumConstants(rop_rates.data());
+    return jac - calculateCompositionDerivatives(m_revProductStoich, rop_rates);
+}
+
+void InterfaceKinetics::setDerivativeSettings(const AnyMap& settings)
+{
+    bool force = settings.empty();
+    if (force || settings.hasKey("skip-coverage-dependence")) {
+        m_jac_skip_cov_dependence = settings.getBool("skip-coverage-dependence",
+            false);
+    }
+    if (force || settings.hasKey("skip-electrochemistry")) {
+        m_jac_skip_electrochem = settings.getBool("skip-electrochemistry",
+            false);
+    }
+    if (force || settings.hasKey("rtol-delta")) {
+        m_jac_rtol_delta = settings.getDouble("rtol-delta", 1e-8);
+    }
+}
+
+void InterfaceKinetics::getDerivativeSettings(AnyMap& settings) const
+{
+    settings["skip-coverage-dependence"] = m_jac_skip_electrochem;
+    settings["skip-electrochemistry"] = m_jac_skip_cov_dependence;
+    settings["rtol-delta"] = m_jac_rtol_delta;
+}
+
+void InterfaceKinetics::processFwdRateCoefficients(double* ropf)
+{
+    // evaluate rate constants and equilibrium constants at temperature and phi
+    // (electric potential)
+    // get updated activities (rates updated below)
+    updateROP();
+    // copy rate coefficients into ropf
+    copy(m_rfn.begin(), m_rfn.end(), ropf);
+
+    // Scale the forward rate coefficient by the perturbation factor
+    for (size_t i = 0; i < nReactions(); ++i) {
+        ropf[i] *= m_perturb[i];
+    }
+}
+
+Eigen::SparseMatrix<double> InterfaceKinetics::calculateCompositionDerivatives(
+    StoichManagerN& stoich, const vector_fp& in)
+{
+    Eigen::SparseMatrix<double> out;
+    vector_fp& outV = m_rbuf1;
+    // derivatives handled by StoichManagerN
+    copy(in.begin(), in.end(), outV.begin());
+    out = stoich.derivatives(m_actConc.data(), outV.data());
+    return out;
+}
+
+void InterfaceKinetics::assertDerivativesValid(const string& name)
+{
+    if (!m_jac_skip_cov_dependence) {
+        throw NotImplementedError(name, "Coverage-dependent reactions not supported.");
+    }
+    else if (!m_jac_skip_electrochem) {
+        throw NotImplementedError(name, "Electrochemical reactions not supported.");
+    }
```

Comment:
This is a bit restrictive -- what if there are no electrochemical or coverage-dependent reactions?

Also, a minor formatting suggestion:
```suggestion
    if (!m_jac_skip_cov_dependence) {
        throw NotImplementedError(name, "Coverage-dependent reactions not supported.");
    } else if (!m_jac_skip_electrochem) {
        throw NotImplementedError(name, "Electrochemical reactions not supported.");
    }
```